zebra-chain/src/types/amount.rs

```diff
@@ -0,0 +1,117 @@
+//!
+use std::{
+    convert::{TryFrom, TryInto},
+    marker::PhantomData,
+    ops::Range,
+};
+
+/// A runtime validated type for representing amounts of zatoshis
+#[derive(Debug, Eq, PartialEq, Clone, Copy, Serialize, Deserialize)]
+pub struct Amount<C = NegativeAllowed>(i64, PhantomData<C>);
+
+impl<C> From<Amount<C>> for i64 {
+    fn from(amount: Amount<C>) -> Self {
+        amount.0
+    }
+}
+
+impl<C> From<Amount<C>> for u64 {
+    fn from(amount: Amount<C>) -> Self {
+        amount.0 as _
+    }
+}
+
+impl<C> TryFrom<i64> for Amount<C>
+where
+    C: AmountConstraint,
+{
+    type Error = Error;
+
+    fn try_from(value: i64) -> Result<Self, Self::Error> {
+        C::validate(value).map(|v| Self(v, PhantomData))
+    }
+}
+
+impl<C> TryFrom<u64> for Amount<C>
+where
+    C: AmountConstraint,
+{
+    type Error = Error;
+
+    fn try_from(value: u64) -> Result<Self, Self::Error> {
+        let value = value.try_into()?;
+
+        C::validate(value).map(|v| Self(v, PhantomData))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use proptest::prelude::*;
+    use std::fmt;
+
+    impl<C> Arbitrary for Amount<C>
+    where
+        C: AmountConstraint + fmt::Debug,
+    {
+        type Parameters = ();
+
+        fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+            C::valid_range().prop_map(|v| Self(v, PhantomData)).boxed()
+        }
+
+        type Strategy = BoxedStrategy<Self>;
+    }
+}
```

Comment:
I think this would be cleaner at the bottom of the file rather than in the middle

Reply:
Is it okay if I leave this with the struct its impled on for now? If this merges before the arbitrary consolidation PR i'll move it as part of that PR and if the  other PR merges first I'll move it in this PR.

Reply:
Yeah sure sounds good